Solutions/Corvus.Monitoring.AspnetCore.Mvc.Specs/Fakes/TestOperationInstance.cs

```diff
@@ -0,0 +1,82 @@
+﻿// <copyright file="TestOperationInstance.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Corvus.Monitoring.AspnetCore.Mvc.Specs.Fakes
+{
+    using System.Collections.ObjectModel;
+    using Corvus.Monitoring.Instrumentation;
+
+    public class TestOperationInstance : IOperationInstance
+    {
+        public TestOperationInstance(string name, AdditionalInstrumentationDetail? additionalDetail = null)
+        {
+            this.Name = name;
+
+            if (additionalDetail?.MetricsIfPresent != null)
+            {
+                foreach (KeyValuePair<string, double> metric in additionalDetail.Metrics)
+                {
+                    this.Metrics.Add(metric);
+                }
+            }
+
+            if (additionalDetail?.PropertiesIfPresent != null)
+            {
+                foreach (KeyValuePair<string, string> property in additionalDetail.Properties)
+                {
+                    this.Properties.Add(property);
+                }
+            }
+        }
+
+        public IDictionary<string, double> Metrics { get; private set; } = new Dictionary<string, double>();
+
+        public IDictionary<string, string> Properties { get; private set; } = new Dictionary<string, string>();
+
+        public string Name { get; }
+
+        public bool IsDisposed { get; private set; } = false;
+
+        public DateTimeOffset StartDateTimeUtc { get; } = DateTimeOffset.UtcNow;
+
+        public DateTimeOffset? EndDateTimeUtc { get; private set; }
+
+        public TimeSpan Duration
+        {
+            get
+            {
+                if (!this.IsDisposed)
+                {
+                    throw new InvalidOperationException("Cannot get the duration of a TestOperationInstance until it has been disposed");
+                }
+
+                return this.EndDateTimeUtc!.Value - this.StartDateTimeUtc;
+            }
+        }
+
+        public void AddOperationMetric(string name, double value)
+        {
+            this.Metrics.Add(name, value);
+        }
+
+        public void AddOperationProperty(string name, string value)
+        {
+            this.Properties.Add(name, value);
+        }
+
+        public void Dispose()
+        {
+            if (this.IsDisposed)
+            {
+                throw new ObjectDisposedException(this.Name);
```

Comment:
By the normal rules of `IDisposable` this shouldn't throw. You are very much allowed to `Dispose` the same object multiple times.

However, as per my earlier comment, I see we aren't exactly using normal `IDisposable` rules here, so if there's a good reason we need to detect double disposal and complain about it, then OK.

Reply:
My expectation is that these `IOperationInstance`s should only be disposed once. However I've had a poke through the source code for Microsoft.ApplicationInsights.Extensibility and I see that the "real" class for App Insights doesn't object to being disposed multiple times.

I'd prefer to leave it as is in the tests though, because that should detect whether our code is disposing the `IOperationInstance` multiple times - which is not the behaviour I'd expect. I'll add some comments in to that effect.